### PR TITLE
Add groundwork for reports

### PR DESCRIPTION
**Pull request by FriedrichRober**
Currently this just provides access to the job status of each package.
We could combine this with the reports script etc. from `gap-infra/integration` at a later point.

### CODE REVIEW

**Comment by FriedrichRober** [comment]
@fingolfin, you can see the action [here](https://github.com/FriedrichRober/PackageDistro/actions/runs/1988061418).

I print the job-status into a json-file in the matrix-job and store it as an artifact.

In the next job I download all reports-artifacts (currently with an unofficial download-artifact).

**Comment by FriedrichRober** [comment]
Is it possible to exclude `primgrp` from packages, since the tests take so long? It seems, that I cannot start GAP / build packages without it.

**Comment by fingolfin** [comment]
Thanks, I'll take a look.

BTW, a less confusing way to limit tests to just a few packages is to modify the loop generating the build matrix, and insert something like
```
              if [[ ${PKG} == browse ]]; then
                break # FIXME: HACK for faster testing, remove before merge
              fi
```
so that it only adds packages up to the given one. Doing that results in a much smaller PR :-)

**Comment by fingolfin** [comment]
OK, I get it, that is indeed not so hard. So this collects the status of each job in the build matrix via artifacts. Clever.

But what it does not yet do is allow us to determine state *changes*. I.e.: these 3 packages FAIL on `main` and now PASS; while these 2 packages PASS on `main` but now `FAIL`.  So overall the number of FAILs decreases, but still some packages newly fail, and we'd want to catch that.

Any idea how to go about that? I guess we could store the aggregated test results (i.e., a list of pairs "packages => pass/fail") for `main` somewhere, and then download it from there and compare. Aha, an idea: we should run `test-all.yml` whenever there is a push to `main` anyway. So we modify `assemble-distro.yml` to call `test-all.yml` (see https://docs.github.com/en/actions/using-workflows/reusing-workflows). A workflow called this way can specify an output, which would be the test results as JSON. We can then upload this as yet another file to the `latest` tag, say `test-status.json`. 

Then in turn any `test-all.yml` run can download `test-status.json` from the `latest` tag (though it should be prepared for it missing), and compare the test status against that.

One has to think a bit about race conditions here. To help debug these, `test-status.json` should contain some extra info, e.g. the date it was created and the SHA1 of the commit on which `test-all.yml` did run.



--

While I think the above should work, for completeness, a completely different approach would be to imitate what https://github.com/gap-actions/should-i-notify-action does...

File: .github/workflows/test-all.yml [hunk 1]
```diff
@@ -200,8 +199,65 @@ jobs:
...
+        with:
+         pattern: reports-*
+         path: _reports
+
+      - name: "Show Files"
+        run: |
+          cd _reports
+          ls
+          cd reports-ace
+          ls
```
Comment (fingolfin) [inline .github/workflows/test-all.yml:263]: So this bit should be turned into an output of the workflow, along the lines of https://docs.github.com/en/actions/using-workflows/reusing-workflows#using-outputs-from-a-reusable-workflow

Then we'll invoke `test-all.yml` from various other workflows which can handle the output as they deem fit

**Comment by FriedrichRober** [comment]
So what I did it in the `gap-infra/integration` repo was to create a `report.md` and to upload each report into a new branch, called `data`. This way we had a complete history of reports. There I had a symlink directory called `latest`, which I used to download the latest report for the comparison.
This branch could look like [this](https://github.com/FriedrichRober/integration/tree/data) and a report could look similar to [this](https://github.com/FriedrichRober/integration/blob/data/reports/by_hash/3b31ecd/report.md).

**Comment by FriedrichRober** [comment]
I also did a little bit more, by creating custom badges and uploading redirects into `gh-pages`, so we could preview a more detailed status and reference these reports inside our `README.md`.

File: .github/workflows/test-all.yml (hunk 1 above)
Comment (FriedrichRober) [reply .github/workflows/test-all.yml:272]: This makes sense. So at this stage we would just collect all data into `test-status.json` with some additional information and turn this into the output. I probably would write a python script for this, so we could test locally with custom stuff in `_reports`.

**Comment by FriedrichRober** [comment]
@fingolfin , how should we handle `status` and `status_only_needed` when counting the total successes or failures? Should we have some sort of hierarchy as `failure` > `cancelled` > `success` and use the maximum for the totals?

**Comment by FriedrichRober** [comment]
@fingolfin, I get a weird error if I run the [action](https://github.com/FriedrichRober/PackageDistro/runs/5569994945?check_suite_focus=true), which I cannot reproduce on my local machine with the same json-files.

`UnicodeDecodeError: 'utf-8' codec can't decode byte 0x8b in position 1: invalid start byte`

Do you know how to fix this?

**Comment by fingolfin** [comment]
There are still some conflicts here, could you please rebase this PR on latest `main`?

**Comment by FriedrichRober** [comment]
@fingolfin, I now did a rebase onto `main`.

**Comment by fingolfin** [comment]
Sorry, no immediate idea on the `UnicodeDecodeError` error. Which JSON files does this happen with? Perhaps some package use PackageInfo.g files that are not valid UTF-8 (I don't think we ever formally specified that `PackageInfo.g` should be in UTF-8, but we defacto expect it to be. Maybe should document it like that, too)

File: .github/workflows/test-all.yml [hunk 2]
```diff
@@ -117,7 +117,7 @@ jobs:
           # "bundled" with NormalizInterface, as this is very slow (takes 5-10
           # minutes). Instead we want it to use libnormaliz-dev installed by
           # us
-          rm -f */build-normaliz.sh
+          rm -f **/build-normaliz.sh
```
Comment (fingolfin) [inline .github/workflows/test-all.yml:120]: Why is this needed?

File: .github/workflows/test-all.yml [hunk 3]
```diff
@@ -206,8 +209,71 @@ jobs:
           PKG=${{ matrix.package }}
           gap -A --quitonbreak -r -c "SetInfoLevel(InfoPackageLoading, PACKAGE_WARNING);res:=TestPackage(\"$PKG\");FORCE_QUIT_GAP(res);"
 
+      - name: "Get job status"
+        id: job-status-default
+        if: ${{ always() }}
+        run: |
+          PKG="${{ matrix.package }}"
```
Comment (fingolfin) [inline .github/workflows/test-all.yml:216]: Why is this line needed?

File: .github/workflows/test-all.yml (hunk 2 above)
Comment (FriedrichRober) [reply .github/workflows/test-all.yml:120]: Probably not needed. I do not know why I thought it was neccessary.

File: .github/workflows/test-all.yml [hunk 4]
```diff
@@ -206,8 +209,71 @@ jobs:
...
         run: |
           PKG=${{ matrix.package }}
           gap -A --quitonbreak -r -c "SetInfoLevel(InfoPackageLoading, PACKAGE_WARNING);res:=TestPackage(\"$PKG\":OnlyNeeded);FORCE_QUIT_GAP(res);"
+
+      - name: "Get job status with OnlyNeeded"
+        id: job-status-only-needed
+        if: ${{ always() }}
+        run: |
+          PKG="${{ matrix.package }}"
+          echo '::set-output name=status::${{ job.status }}'
```
[the remark this answers is not in the document]
Comment (fingolfin) [reply .github/workflows/test-all.yml:231]: Alternatively, [according to the docs](https://docs.github.com/en/actions/learn-github-actions/contexts#steps-context) you can use `steps.<step_id>.outcome` to get the state of a specific step. But then you can do this right at the end, no need to have two steps collecting the status?

File: .github/workflows/test-all.yml [hunk 5]
```diff
@@ -206,8 +209,71 @@ jobs:
...
+
+      - name: "Create job status json-file"
+        if: ${{ always() }}
+        run: |
+          PKG="${{ matrix.package }}"
+          STATUS_DEFAULT="${{ steps.job-status-default.outputs.status }}"
+          STATUS_ONLY_NEEDED="${{ steps.job-status-only-needed.outputs.status }}"
```
Comment (fingolfin) [inline .github/workflows/test-all.yml:238]: To my understanding, these two variables will end up with string values like `success`, `failure`, `cancelled`, or `skipped`, right? Perhaps we should add a comment here stating that, possibly with a link to the relevant docs.

File: .github/workflows/test-all.yml [hunk 6]
```diff
@@ -206,8 +209,71 @@ jobs:
...
+          jq --null-input \
+            --arg status_default "${STATUS_DEFAULT}" \
+            --arg status_only_needed "${STATUS_ONLY_NEEDED}" \
+            '{"status_default": $status_default, "status_only_needed": $status_only_needed}' > ${PKG}.json
```
Comment (fingolfin) [inline .github/workflows/test-all.yml:242]: What does using `jq` here gain us over just
```suggestion
          echo "{\"status_default\": \"$STATUS_DEFAULT\", \"status_only_needed\": \"$STATUS_ONLY_NEEDED\"}" > ${PKG}.json
```

**Comment by FriedrichRober** [comment]
> 



> Sorry, no immediate idea on the `UnicodeDecodeError` error. Which JSON files does this happen with? Perhaps some package use PackageInfo.g files that are not valid UTF-8 (I don't think we ever formally specified that `PackageInfo.g` should be in UTF-8, but we defacto expect it to be. Maybe should document it like that, too)

It happens with the json files we generate via `jq`, so the job-status we upload via artifacts. The weird thing is, that after downloading the artifacts manually and trying the script locally on my laptop, everything works fine. I now try the option to ignore unknown encodings in the script. Let's see if this works.

File: .github/workflows/test-all.yml (hunk 3 above)
Comment (FriedrichRober) [reply .github/workflows/test-all.yml]: Removed. This is not needed anymore.

File: .github/workflows/test-all.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (FriedrichRober) [reply .github/workflows/test-all.yml:229]: You are right. I will try this.

File: .github/workflows/test-all.yml (hunk 6 above)
Comment (FriedrichRober) [reply .github/workflows/test-all.yml:240]: I initially planned to add more data and thought it was easier to use jq here, so I didn't have to worry about escaping special characters. But now, this is overkill. I commit your suggestion.

File: .github/workflows/test-all.yml (hunk 5 above)
Comment (FriedrichRober) [reply .github/workflows/test-all.yml:236]: I will add a comment there and a link to the documentation. With `steps.<step_id>.outcome` we get four states, with `job.status` we get three states, so I will need to adjust the scripts slightly.

File: .github/workflows/test-all.yml [hunk 7]
```diff
@@ -201,13 +204,69 @@ jobs:
...
+        uses: actions/upload-artifact@v2
+        with:
+          name: "reports-${{ matrix.package }}"
+          path: "${{ matrix.package }}.json"
+
+  report:
+    name: "Report"
+    needs: test-package
+    if: ${{ always() }}
+    runs-on: ubuntu-18.04
```
Comment (fingolfin) [inline .github/workflows/test-all.yml:243]: Do we need to use an old Ubuntu for some reason? Then we should say so in a comment. Otherwise:
```suggestion
    runs-on: ubuntu-latest
```

Comment (FriedrichRober) [reply .github/workflows/test-all.yml:243]: This is probably copy-paste-pasta from some other old actions we used.

**Comment by FriedrichRober** [comment]
Ok, the pull-request is now nearly finished, except for the broken workflow run id `${{ github.run_id }}`. The same line of code worked in `gap-infra/integration`.

File: .github/workflows/test-all.yml (hunk 1 above)
Comment (FriedrichRober) [reply .github/workflows/test-all.yml]: Ok, I turned this now into an output for workflow calls.

**Comment by fingolfin** [comment]
There are several packages which use LATIN1 encoding instead of UTF-8 in their PackageInfo.g, and these cause problems. I'll try to fix as many as I can

**Comment by fingolfin** [comment]
Thank you! I'll have a look. And sorry, but it seems you'll have to rebase once more 😬

**Comment by FriedrichRober** [comment]
@fingolfin, I think you can squash and merge this. Here is the [action](https://github.com/FriedrichRober/PackageDistro/actions/runs/1997993269) (before rebase) with all tests enabled.

Question: Why did you remove the input for the workflow_call?

I will do the comparison report in the following pull request.

**Comment by FriedrichRober** [comment]
Here is the [action](https://github.com/FriedrichRober/PackageDistro/runs/5585212961?check_suite_focus=true) after rebase. Seems to work correctly.

**Review by fingolfin — APPROVED** [review]
Looks good to me, great!!!

**Comment by fingolfin** [comment]
Ah damn I should have squashed. Ah well, whatever (unless it annoys you, then I can rewrite history now)
